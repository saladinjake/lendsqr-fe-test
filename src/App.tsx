import React from 'react';
import { Route, Routes } from 'react-router-dom';

function App() {
	return (
		<div>
			<Routes>
				<Route path="/" element={<>Login</>} />
				<Route path="/login" element={<>Login</>} />
				<Route path="/dashboard/*" element={<>Dashboard</>}>
					<Route index element={<>Users</>} />
					<Route index path="users" element={<>Users page</>} />
					<Route path="users/:id/profile" element={<>user details</>} />
				</Route>
				<Route path="*" element={<h1>PAGE NOT FOUND</h1>} />
			</Routes>
		</div>
	);
}

export default App;
