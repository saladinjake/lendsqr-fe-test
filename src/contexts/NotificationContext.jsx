import React, { useState } from "react";
const NotificationContext = React.createContext({
  notification: null,
  notificationMessage: null,
  success: (text)=> {},
  error: (text) => {},
  clear: ()=> {},
});

const STATES = {
  SUCCESS: "success",
  ERROR: "error",
};

const NotificationProvider = (props) => {
  const [notification, setNotification] = useState(null);
  const [notificationMessage, setNotificationText] = useState(null);
  const success = (text) => {
    window.scroll(0, 0);
    setNotificationText(text);
    setNotification(STATES.SUCCESS);
  };
  const error = (text) => {
    window.scroll(0, 0);
    setNotificationText(text);
    setNotification(STATES.ERROR);
  };
  const clear = () => {
    setNotificationText(null);
    setNotification(null);
  };
  return (
    <NotificationContext.Provider
      value={{
        success,
        error,
        clear,
        notification,
        notificationMessage,
      }}
    >
      {props.children}
    </NotificationContext.Provider>
  );
};

export { NotificationProvider };
export default NotificationContext;
