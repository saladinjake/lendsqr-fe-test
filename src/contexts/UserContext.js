import React, { createContext, useReducer } from "react";
import  ProductReducer  from "../redux/reducers/courses.reducer";



export const ProductsContext = createContext();

const OnlineCoursesContextProvider = ({ children }) => {
  const courses = [] //get all products from api
  const [state] = useReducer(ProductReducer, courses);

  return (
    <ProductsContext.Provider value={[...state]}>
      {children}
    </ProductsContext.Provider>
  );
};

export default OnlineCoursesContextProvider;
