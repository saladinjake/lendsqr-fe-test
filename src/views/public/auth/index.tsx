import { Route, Routes } from "react-router-dom";
import LoginLanding from "../../../modules/auth/login/LoginForm";


function AuthenticationRouter() {
  return (
    <Routes>
      <Route index path="/" element={<LoginLanding />} />
      <Route path="/login" element={<LoginLanding />} />
    </Routes>
  );
}

export default AuthenticationRouter;
