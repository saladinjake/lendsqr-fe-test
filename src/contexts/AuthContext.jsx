import React, {
  useReducer,
  createContext,
  useContext,
  useState,
  useEffect,
} from "react";
import FingerprintJS from "@fingerprintjs/fingerprintjs";

import * as authService from "../api/auth.service";

import NotificationContext from "./NotificationContext";
import { connect, useDispatch, useSelector } from "react-redux";
import {
  getAcessTokenFromStore,
  getIsAuthenticatedFromStore,
  getUserRolesFromStore,
} from "../redux/reducers/auth.reducer";

import {
  login as loginSuccessfulDispatcher,
  logOut as cleanUserSession,
} from "../redux/actions/auth.action";

import {
  finishAuth,
  deleteToken,
  isSessionExpired,
  getToken,
  getUserData,
  getUserProfile,
  isTokenStillValid,
} from "../utils/tokenConfig";

import { isAuthenticatedByRoles } from "../utils";

const cachedUser = localStorage && JSON?.parse(localStorage.getItem("user"));
const cachedRole =
  localStorage && JSON?.parse(localStorage.getItem("user_roles"));
const cachedToken = localStorage && localStorage.getItem("token");
const cachedAuthorization = localStorage.user && localStorage.token;

const initialState = {
  isAuthenticated: cachedAuthorization ? true : false,
  user: cachedUser ? cachedUser : null,
  token: cachedToken ? cachedToken : null,
  user_roles: cachedRole ? cachedRole : null,
  isAuth: cachedAuthorization ? true : false, //getToken() ? true : false,
  isLoading: false,
  userProfile: null,
  isLoggedIn: false,
  isLoginPending: false,
  loginError: null,
  loadAuthUser: () => isTokenStillValid(),
};

const AuthContext = createContext({
  isAuthenticated: cachedAuthorization ? true : false,
  user: cachedUser ? cachedUser : null,
  token: cachedToken ? cachedToken : null,
  user_roles: cachedRole ? cachedRole : null,
  isAuth: cachedAuthorization ? true : false, //getToken() ? true : false,

  isTokenStillValid: () => isTokenStillValid(),
  isLoading: false,
  userProfile: null,
  login: (credentials) => authService.loginUser(credentials),
  logout: () => {},
  forgetPassword: (payload) => {
   //todo
  },
  changePassword: (payload) => {
    //todo
  },
  updatePassword: (payload) => {
     //todo
  },
  loadAuthUser: () => {},
  modules: [],
  permissions: [],
});

function authReducer(state, action) {
  switch (action.type) {
    case "LOGIN_START":
      return {
        ...state,
        isLoading: true,
      };
    case "LOGIN_SUCCESS":
      return {
        ...state,
        user: action.user,
        isAuth: action.isAuthenticated,
        isLoading: false,
        userProfile: action.userProfile,
        user_roles: action.user_roles,
        isAuthenticated: action.isAuthenticated,
      };
    case "LOGOUT":
      return {
        ...state,
        user: null,
        isAuth: false,
        isLoading: false,
        userProfile: null,
      };
    case "FORGOT_PASSWORD_SUCCESS":
      return {
        ...state,
      };
    case "FORGOT_PASSWORD_FAILED":
      cleanUserSession();
      return {
        ...state,
      };
    case "CHANGE_PASSWORD_FAILED":
      cleanUserSession();
      return {
        ...state,
      };
    case "CHANGE_PASSWORD_SUCCESS":
      cleanUserSession();
      return {
        ...state,
      };

    default:
      return state;
  }
}

function getPublicIp(url) {
  return fetch(url).then((res) => res.text());
}

async function loadAuthUser(state, dispatch, logout) {
  try {
    const isUserStillAuthenticated = await state.loadAuthUser(); //check token profile
    const ssoUserIdentityProfile = isUserStillAuthenticated?.data?.user;
    const ssoUserRole = isUserStillAuthenticated?.data?.user?.roles;
    const adminOnly =
      Array.isArray(ssoUserRole) &&
      ssoUserRole?.every((role) => role?.name === "User");

    if (!isUserStillAuthenticated) {
      logout();
      await cleanUserSession();
    } else if (ssoUserIdentityProfile && adminOnly) {
      if (!isSessionExpired()) {
        let token = getToken();
        let user = getUserData();
        let userProfile = getUserProfile();
        if (token && user) {
          dispatch({
            type: "LOGIN_SUCCESS",
            user,
            isAuth: true,
            userProfile,
            isAuthenticated: true,
          });
        } else {
          // logout();
          await cleanUserSession();
        }
      } else {
        // logout();
        await cleanUserSession();
      }
    }
  } catch (error) {
    console.log(error);
    // logout();
    await cleanUserSession();
  }
}

function AuthProvider(props) {
  const reduxAwareDispatcher = useDispatch();
  const [state, dispatch] = useReducer(authReducer, initialState);
  const notificationCtx = useContext(NotificationContext);
  const [userIp, setMyIp] = useState(null);
  const [message, setMessage] = useState("");
  const isAuthenticatedFromStore = useSelector((state) =>
    getIsAuthenticatedFromStore(state)
  );
  const getPersistenceUser = useSelector((state) =>
    getUserRolesFromStore(state)
  );
  const getPersistenceRoles = useSelector((state) =>
    getUserRolesFromStore(state)
  );
  const getPersistenceAcessTokenFromStore = useSelector((state) =>
    getAcessTokenFromStore(state)
  );
  const dispatchPersistenceLogin = (userData) =>
    reduxAwareDispatcher(loginSuccessfulDispatcher(userData));
  const dispatchPersistenceLogout = () =>
    reduxAwareDispatcher({ type: "LOGIN_FAIL" });

  const [internalsState, setState] = useState(initialState);
  useEffect(() => {
    loadAuthUser(state, dispatch, logout);
  }, [state]);

  const login = async (payload) => {
    try {
      dispatch({ type: "LOGIN_START" });
      const ipaddr = await getPublicIp(
        "https://www.cloudflare.com/cdn-cgi/trace"
      );
      let ipRegex = /[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}/;
      let ip = ipaddr.match(ipRegex)[0];
      const MaxTimeForRetrials = 3;
      let count = 0;
      const apiResponseData = await authService.loginUser(payload);

      console.log(apiResponseData, ">>>");
      const decoded = apiResponseData?.data;
      const adminRoles = apiResponseData?.data?.user?.roles;
      const hasAdminRoles = isAuthenticatedByRoles("User", getPersistenceRoles);
      if (hasAdminRoles) {
        const userToken = apiResponseData?.data?.access_token;
        const userProfile = {
          id: apiResponseData?.data?.user?.id,
          employeeNumber: apiResponseData?.data?.user?.employeeNumber,
          username: apiResponseData?.data?.user?.username,
          first_name: apiResponseData?.data?.user?.first_name,
          middle_name: apiResponseData?.data?.user?.middle_name,
          last_name: apiResponseData?.data?.user?.last_name,
          email: apiResponseData?.data?.user?.email,
          phone_number: apiResponseData?.data?.user?.phone_number,
          image_url: apiResponseData?.data?.user?.image_url,
          instructor_profile: apiResponseData?.data?.user?.instructor_profile,
          roles: apiResponseData?.data?.user?.roles,
        };

        const tokenExpiry = ""; //decoded?.expires_at || null;
        const fpPromise = await FingerprintJS.load();
        const fpPromiseGetResult = await fpPromise.get();
        const visitorId = fpPromiseGetResult.visitorId;

        const data = {
          userToken,
          auditLogData: {
            SYSTEMIDENTIFIER: "",
            SYSTEMIPADDRESS: ip,
            SYSTEMMACADDRESS: visitorId,
            SYSTEMNAME: "",
            SUBJECTIDENTIFIER: "",
          },
        };

        try {
          const user = apiResponseData.data?.user;
          const ApiResponseStatus =
            apiResponseData?.data?.success &&
            apiResponseData?.data?.status_code;

          if (ApiResponseStatus) {
            finishAuth(userToken, user, visitorId, tokenExpiry, userProfile);
            dispatchPersistenceLogin({
              user,
              isAuthenticated: true,
              access_token: userToken,
              user_roles: getPersistenceRoles,
            });
            notificationCtx.success("Login successful!");
            dispatch({
              type: "LOGIN_SUCCESS",
              user,
              isAuth: isAuthenticatedFromStore, //true,
              userProfile,
              user_roles: getPersistenceRoles,
            });
          } else {
            notificationCtx.error("Invalid credentials. Please Try again");
          }
        } catch (error) {
          count = count + 1;
          dispatch({ type: "LOGOUT" });
          dispatchPersistenceLogout();
          await cleanUserSession();
          notificationCtx.error("Invalid credentials. Please Try again");
        }
        return apiResponseData.data;
      }
      return apiResponseData.data;
    } catch (error) {
      console.log(error);
      notificationCtx.error("Invalid credentials.");
    }
  };

  async function logout() {
    deleteToken();
    dispatch({ type: "LOGOUT" });
    dispatchPersistenceLogout();
    await cleanUserSession();
  }

  async function forgetPassword(payload) {
  
  }

  async function changePassword(payload) {
    
  }

  async function updatePassword(payload) {
    
  }

  return (
    <AuthContext.Provider
      value={{
        user: state.user,
        login,
        logout,
        forgetPassword,
        changePassword,
        updatePassword,
        isAuth: state.isAuth,
        isLoading: state.isLoading,
        loadAuthUser: () => loadAuthUser(state, dispatch, logout),
        userProfile: state.userProfile,
      }}
      {...props}
    >
      {props.children}
    </AuthContext.Provider>
  );
}

const useAuth = () => useContext(AuthContext);
export { AuthContext, AuthProvider, useAuth };
