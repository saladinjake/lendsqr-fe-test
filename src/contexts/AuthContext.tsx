import React, {
  useReducer,
  createContext,
  useContext,
  useState,
  useEffect,
} from "react";
import FingerprintJS from "@fingerprintjs/fingerprintjs";

import * as authService from "../api/auth.service";

import NotificationContext from "./NotificationContext";
import { connect, useDispatch, useSelector } from "react-redux";
import {
  getAcessTokenFromStore,
  getIsAuthenticatedFromStore,
  getUserRolesFromStore,
} from "../redux/reducers/auth.reducer";

import {
  login as loginSuccessfulDispatcher,
  logOut as cleanUserSession,
} from "../redux/actions/auth.action";

import {
  finishAuth,
  deleteToken,
  isSessionExpired,
  getToken,
  getUserData,
  getUserProfile,
  isTokenStillValid,
} from "../utils/tokenConfig";

import { isAuthenticatedByRoles } from "../utils";
import { RootState } from "../redux/reducers";
const cachedUser = localStorage && JSON?.parse(""+localStorage.getItem("user"));
const cachedRole =
  localStorage && JSON?.parse(""+ localStorage.getItem("user_roles"));
const cachedToken = localStorage && localStorage.getItem("token");
const cachedAuthorization = localStorage.user && localStorage.token;

const initialState = {
  isAuthenticated: cachedAuthorization ? true : false,
  user: cachedUser ? cachedUser : null,
  token: cachedToken ? cachedToken : null,
  user_roles: cachedRole ? cachedRole : null,
  isAuth: cachedAuthorization ? true : false, //getToken() ? true : false,
  isLoading: false,
  userProfile: null,
  isLoggedIn: false,
  isLoginPending: false,
  loginError: null,
  loadAuthUser: () => isTokenStillValid(),
};

interface IDecoded {
  email: string;
  family_name: string;
  given_name: string;
  name: string;
  picture: string;
  exp: string;
  designation?: string;
  employeeNumber?: string;
  firstName?: string;
  lastName?: string;
  userRoleID?: string;
  id?: string;
  userName?: string;
  password?: string;
  isSuccessful?: boolean;
}

const AuthContext = createContext({
  isAuthenticated: cachedAuthorization ? true : false,
  user: cachedUser ? cachedUser : null,
  token: cachedToken ? cachedToken : null,
  user_roles: cachedRole ? cachedRole : null,
  isAuth: cachedAuthorization ? true : false, //getToken() ? true : false,

  isTokenStillValid: () => isTokenStillValid(),
  isLoading: false,
  userProfile: null,
  login: (credentials: any) => authService.loginUser(credentials),
  logout: () => {},
  forgetPassword: (payload: any): any =>  {
    
  },
  changePassword: (payload: any): any =>  {
   
  },
  updatePassword: (payload: any): any =>{
   
  },
  loadAuthUser: () => {},
  modules: [],
  permissions: [],
});

function authReducer(state:any, action:any) {
  switch (action.type) {
    case "LOGIN_START":
      return {
        ...state,
        isLoading: true,
      };
    case "LOGIN_SUCCESS":
      return {
        ...state,
        user: action.user,
        isAuth: action.isAuthenticated,
        isLoading: false,
        userProfile: action.userProfile,
        user_roles: action.user_roles,
        isAuthenticated: action.isAuthenticated,
      };
    case "LOGOUT":
      return {
        ...state,
        user: null,
        isAuth: false,
        isLoading: false,
        userProfile: null,
      };
    case "FORGOT_PASSWORD_SUCCESS":
      return {
        ...state,
      };
    case "FORGOT_PASSWORD_FAILED":
      // logOut();
      return {
        ...state,
      };
    case "CHANGE_PASSWORD_FAILED":
      // logOut();
      return {
        ...state,
      };
    case "CHANGE_PASSWORD_SUCCESS":
      // logOut();
      return {
        ...state,
      };

    default:
      return state;
  }
}



async function loadAuthUser(state: any, dispatch: any, logout: any) {

  try {
    const isUserStillAuthenticated = await state.loadAuthUser(); //check token profile
    const ssoUserIdentityProfile = isUserStillAuthenticated?.data?.user;
    const ssoUserRole = isUserStillAuthenticated?.data?.user?.roles;
    const adminOnly = Array.isArray(ssoUserRole) && ssoUserRole?.every((role) => role?.name === "SuperAdmin");

    if (!isUserStillAuthenticated) {
      // logOut();
      await cleanUserSession();
    } else if (ssoUserIdentityProfile && adminOnly) {
      if (!isSessionExpired()) {
        let token = getToken();
        let user = getUserData();
        let userProfile = getUserProfile();
        if (token && user) {
          dispatch({
            type: "LOGIN_SUCCESS",
            user,
            isAuth: true,
            userProfile,
            isAuthenticated: true,
          });
        } else {
          // logout();
          await cleanUserSession();
        }
      } else {
        // logout();
        await cleanUserSession();
      }
    }
  } catch (error) {
    console.log(error);
    // logout();
    await cleanUserSession();
  }
}

function AuthProvider(props: any) {
  const reduxAwareDispatcher = useDispatch();
  const [state, dispatch] = useReducer(authReducer, initialState);
  const notificationCtx = useContext(NotificationContext);
  const [userIp, setMyIp] = useState(null);
  const [message, setMessage] = useState("");
  const isAuthenticatedFromStore = useSelector((state: RootState) =>
    getIsAuthenticatedFromStore(state)
  );
  const getPersistenceUser = useSelector((state: RootState) =>
    getUserRolesFromStore(state)
  );
  const getPersistenceRoles = useSelector((state: RootState) =>
    getUserRolesFromStore(state)
  );
  const getPersistenceAcessTokenFromStore = useSelector((state: RootState) =>
    getAcessTokenFromStore(state)
  );
  const dispatchPersistenceLogin = (userData: any) =>
    reduxAwareDispatcher(loginSuccessfulDispatcher(userData) as any);
  const dispatchPersistenceLogout = () =>
    reduxAwareDispatcher({ type: "LOGIN_FAIL" });

  const [internalsState, setState] = useState(initialState);
  useEffect(() => {
    loadAuthUser(state, dispatch, logout);
  }, [state]);

  const login = async (payload: any) => {
    try {
      dispatch({ type: "LOGIN_START" });
    
      const MaxTimeForRetrials = 3;
      let count = 0;
      const apiResponseData : any= await authService.loginUser(payload);
      const decoded = apiResponseData?.data;
      const adminRoles = apiResponseData?.data?.user?.roles;
      const hasAdminRoles = isAuthenticatedByRoles(
        "SuperAdmin",
        getPersistenceRoles
      );
      if (hasAdminRoles) {
        const userToken = apiResponseData?.data?.access_token;
        const userProfile = {
          id: apiResponseData?.data?.user?.id,
          employeeNumber: apiResponseData?.data?.user?.employeeNumber,
          username: apiResponseData?.data?.user?.username,
          first_name: apiResponseData?.data?.user?.first_name,
          middle_name: apiResponseData?.data?.user?.middle_name,
          last_name: apiResponseData?.data?.user?.last_name,
          email: apiResponseData?.data?.user?.email,
          phone_number: apiResponseData?.data?.user?.phone_number,
          image_url: apiResponseData?.data?.user?.image_url,
          instructor_profile: apiResponseData?.data?.user?.instructor_profile,
          roles: apiResponseData?.data?.user?.roles,
        };
     
        const tokenExpiry = ""; //decoded?.expires_at || null;
        const fpPromise = await FingerprintJS.load();
        const fpPromiseGetResult = await fpPromise.get();
        const visitorId = fpPromiseGetResult.visitorId;

        const data = {
          userToken,
          auditLogData: {
            SYSTEMIDENTIFIER: "",
            SYSTEMIPADDRESS: "192.168.1.1",
            SYSTEMMACADDRESS: visitorId,
            SYSTEMNAME: "",
            SUBJECTIDENTIFIER: "",
          },
        };

        try {
          const user = apiResponseData.data?.user;
          const ApiResponseStatus =
            apiResponseData?.data?.success &&
            apiResponseData?.data?.status_code;

          if (ApiResponseStatus) {
            finishAuth(userToken, user, visitorId, tokenExpiry, userProfile);
            dispatchPersistenceLogin({
              user,
              isAuthenticated: true,
              access_token: userToken,
              user_roles: getPersistenceRoles,
            });
            notificationCtx.success("Login successful!");
            dispatch({
              type: "LOGIN_SUCCESS",
              user,
              isAuth: isAuthenticatedFromStore, //true,
              userProfile,
              user_roles: getPersistenceRoles,
            });
          } else {
            notificationCtx.error("Invalid credentials. Please Try again");
          }
        } catch (error) {
          count = count + 1;
          dispatch({ type: "LOGOUT" });
          dispatchPersistenceLogout();
          await cleanUserSession();
          notificationCtx.error("Invalid credentials. Please Try again");
        }
        return apiResponseData.data;
      }
      return apiResponseData.data;
    } catch (error) {
      console.log(error);
      notificationCtx.error("Invalid credentials.");
    }
  };

  async function logout() {
    deleteToken();
    dispatch({ type: "LOGOUT" });
    dispatchPersistenceLogout();
    await cleanUserSession();
  }

  async function forgetPassword(payload:any) {
   
  }

  async function changePassword(payload: any) {
   
  }

  async function updatePassword(payload: any) {
    
  }

  return (
    <AuthContext.Provider
      value={{
        user: state.user,
        login,
        logout,
        forgetPassword,
        changePassword,
        updatePassword,
        isAuth: state.isAuth,
        isLoading: state.isLoading,
        loadAuthUser: () => loadAuthUser(state, dispatch, logout),
        userProfile: state.userProfile,
      }}
      {...props}
    >
      {props.children}
    </AuthContext.Provider>
  );
}

const useAuth = () => useContext(AuthContext);
export { AuthContext, AuthProvider, useAuth };