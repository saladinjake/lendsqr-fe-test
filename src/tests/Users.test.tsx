import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import App from "../routes/index";

import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { RouterProvider } from "react-router-dom";
import { enableFetchMocks } from "jest-fetch-mock";

import { Provider as Provisioner } from "react-redux";
import Store from "../redux/store";
import { NotificationProvider } from "../contexts/NotificationContext";
import NotificationBar from "../components/shared/Notifier";
//import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
import AppNavigator from "../routes";
import { AuthProvider } from "../contexts/AuthContext";
import { loginUser, logoutUser } from "../api/auth.service";
import { getAuthProfile } from "../api/user.service";
import "@testing-library/jest-dom";
import { rest } from "msw";
import { setupServer } from "msw/node";
import UsersDashBoard from "../modules/dashboard/index";



import { BrowserRouter } from "react-router-dom";
const modeUrl = process.env.NODE_ENV == "development" ? "http://localhost:3000" : "https://victorjuwa-lendsqr-fe-test.vercel.app";

const mockServer = setupServer(
  rest.get(`${modeUrl}/users`, (req, res, ctx) => {
    return res(
      ctx.json({
        firstName: "victor",
        lastName: "Juwa",
        role: "Lead developer",
      })
    );
  })
);

beforeAll(() => mockServer.listen());
afterEach(() => mockServer.resetHandlers());
afterAll(() => mockServer.close());
const authClient = {
    isAuthenticated: getAuthProfile,
    login: loginUser,
    logout: logoutUser,
  };

describe("Users lists", () => {
    const initialState = { output: 10 };
    const mockStore = configureStore();
    let store;
  it("renders without crashing", () => {
     const component = render(
        <React.StrictMode>
        <Provisioner store={Store}>
          <AuthProvider client={authClient}>
            <UsersDashBoard />
          </AuthProvider>
        </Provisioner>
      </React.StrictMode>
    );
    expect(component.container).toMatchSnapshot();

    expect(screen.getByText(/102,453/i)).toBeInTheDocument();
    expect(screen.getByText(/Active Users/i)).toBeInTheDocument();
    expect(screen.getByText(/Users with loans/i)).toBeInTheDocument();
    expect(screen.getByText(/Users with savings/i)).toBeInTheDocument();
  });

  it('should display list of users', async () => {
  	 const { container } = render(  <React.StrictMode>
        <Provisioner store={Store}>
          <AuthProvider client={authClient}>
            <UsersDashBoard />
          </AuthProvider>
        </Provisioner>
      </React.StrictMode>
  	);

      
  	waitFor(() => {
  		const table = screen.getByTestId('table-responsive');
    
  		expect(table).toBeInTheDocument();


          let usersList: any = container.querySelector('.random-selector-test-case') ;
          expect(usersList.length).toBeGreaterThan(0);

          usersList.forEach( (item:any, index: number)  => {
             expect(item).toHaveAttribute('class', index +'-test-users')
           })
  	});
  });
});
