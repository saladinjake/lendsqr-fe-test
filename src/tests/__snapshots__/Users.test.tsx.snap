// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users lists renders without crashing 1`] = `
<div>
  <div
    class="dashboard-module"
  >
    <h1>
      Users
    </h1>
    <div
      class="data-statistics"
      data-testid="statistics-data"
      id="statistics"
    >
      <div
        class="statistics-info-detail"
      >
        <img
          alt="stat"
          src="/images/icons/user-stat-icon-1.svg"
        />
        <p
          data-testid="statistics-data-0"
        >
          Users
        </p>
        <p>
          2,453
        </p>
      </div>
      <div
        class="statistics-info-detail"
      >
        <img
          alt="stat"
          src="/images/icons/user-stat-icon-2.svg"
        />
        <p
          data-testid="statistics-data-1"
        >
          Active Users
        </p>
        <p>
          2,453
        </p>
      </div>
      <div
        class="statistics-info-detail"
      >
        <img
          alt="stat"
          src="/images/icons/user-stat-icon-3.svg"
        />
        <p
          data-testid="statistics-data-2"
        >
          Users with loans
        </p>
        <p>
          12,453
        </p>
      </div>
      <div
        class="statistics-info-detail"
      >
        <img
          alt="stat"
          src="/images/icons/user-stat-icon-4.svg"
        />
        <p
          data-testid="statistics-data-3"
        >
          Users with savings
        </p>
        <p>
          102,453
        </p>
      </div>
    </div>
    <div
      class="loader"
      style="height: 20px;"
    >
      <svg
        enable-background="new 0 0 0 0"
        id="L9"
        style="height: 90px;"
        version="1.1"
        viewBox="0 0 100 100"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <path
          d="M73,50c0-12.7-10.3-23-23-23S27,37.3,27,50 M30.9,50c0-10.5,8.5-19.1,19.1-19.1S69.1,39.5,69.1,50"
          fill="grey"
        >
          <animatetransform
            attributeName="transform"
            attributeType="XML"
            dur="1s"
            from="0 50 50"
            repeatCount="indefinite"
            to="360 50 50"
            type="rotate"
          />
        </path>
      </svg>
    </div>
    <div
      class="pagination-widget"
    >
      <div
        class="pagination-info"
      >
        <p>
          Showing
           
          <span>
            9
             
            <img
              alt="down arrow"
              src="/images/icons/down-arrow.svg"
            />
          </span>
           
          out of 
          0
        </p>
      </div>
      <ul
        aria-label="Pagination"
        class="pagination-wrapper"
        role="navigation"
      >
        <li
          class="arrow disabled"
        >
          <a
            aria-disabled="true"
            aria-label="Previous page"
            class=" "
            rel="prev"
            role="button"
            tabindex="-1"
          >
            &lt;
          </a>
        </li>
        <li
          class="arrow"
        >
          <a
            aria-disabled="false"
            aria-label="Next page"
            class=""
            rel="next"
            role="button"
            tabindex="0"
          >
            &gt;
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
